### PR TITLE
Add support for storing non-CLP-encodable values in a separate column; Replace CLP row value that are too large to store in FixedByteMVMutableForwardIndex with an error message.

### PR DESCRIPTION
**Pull request by jackluo923**
The dictVar and encodedVar columns in CLP are stored in multi-value forward indexes, which require users to specify the maximum number of values per row at index creation time. However, this is impractical because the number of values per row is only determined during ingestion. To prevent ingestion errors when the number of values in a row exceeds the allowed limit for a multi-value row, the encoded value is replaced with an error message. To preserve the original content, the raw value is stored in a separate field.

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java [hunk 1]
```diff
@@ -525,6 +526,9 @@ public boolean index(GenericRow row, @Nullable RowMetadata rowMetadata)
       }
     }
 
+    // NOTE: we msut do this before we index a single column to avoid partially indexing the row
```
Comment (deemoliu) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:529]: this change seems affecting non-CLP logic, can you please eloborate why this change is needed.

Comment (jackluo923) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:529]: It's an edge case that wasn't handled in the original Pinot code. CLP triggered this edge case in very extreme case. Specifically, if a MV column's content is larger than a chunk, Pinot will error with or without CLP.

Comment (deemoliu) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:529]: @Jackie-Jiang can you help review this section. is this introduce backward incompatible behavior?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java [hunk 2]
```diff
@@ -525,6 +526,9 @@ public boolean index(GenericRow row, @Nullable RowMetadata rowMetadata)
       }
     }
 
+    // NOTE: we must do this before we index a single column to avoid partially indexing the row
```
Comment (chenboat) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:529]: Can we add link to the current hard limit?

**Review by deemoliu — APPROVED** [review]

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java (hunk 2 above)
Comment (deemoliu) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java]: @jackluo923 can you add a unit test to cover this function/scenario?

Comment (jackluo923) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java]: @deemoliu After inspecting the code and writing a unit test, it seems the problem is not reproducible in a unit test because Java is running in `non-production` mode (i.e. `java -ea xxx`) with assertions enabled. This validation check is pretty much equivalent to the assertion found in `org.apache.pinot.segment.local.realtime.impl.forward.FixedByteMVMutableForwardIndex#updateHeader`, except we now throw an exception at run-time.:

```
  private int updateHeader(int row, int numValues) {
    assert (numValues <= _maxNumberOfMultiValuesPerRow);
    int newStartIndex = _prevRowStartIndex + _prevRowLength;
    if (newStartIndex + numValues > _currentCapacity) {
      addDataBuffer(_incrementalCapacity);
      _prevRowStartIndex = 0;
      _prevRowLength = 0;
      newStartIndex = 0;
    }
    writeIntoHeader(row, _dataWriters.size() - 1, newStartIndex, numValues);
    _prevRowStartIndex = newStartIndex;
    _prevRowLength = numValues;
    return newStartIndex;
  }
```

In production-mode, the assert is an no-op and if a value inserted into the MV forward index is greater than the maximum, we get inconsistencies later on in the code. In `validateLengthOfMVColumns`, we use a more relaxed check which validate against `_incrementalCapacity`. FixedByteMVMutableForwardIndex derives the maximum # of values per entry using `_maxNumberOfMultiValuesPerRow`, but `_incrementalCapacity` is the real maximum. See code in `org.apache.pinot.segment.local.realtime.impl.forward.FixedByteMVMutableForwardIndex#FixedByteMVMutableForwardIndex` for more detail.

Comment (jackluo923) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java]: @chenboat  
The current maximum is 1000 (default) entries for MV forward indexes because the code paths which invokes the constructor almost always use the default value. See `org.apache.pinot.segment.local.segment.index.forward.ForwardIndexType#MAX_MULTI_VALUES_PER_ROW` for the default value of 1000. See `org.apache.pinot.segment.local.segment.index.forward.ForwardIndexType#createMutableIndex` for the 2 places where MV mutable forward indexes is created with this 1000 max value limit.
